src/data-connect/data-connect-api-client-internal.ts

```diff
@@ -198,6 +198,203 @@ export class DataConnectApiClient {
+   */
+  private objectToString(data: any): string {
+    if (typeof data === 'string') {
+      const escapedString = data
+        .replace(/\\/g, '\\\\') // Replace \ with \\
+        .replace(/"/g, '\\"');  // Replace " with \"
+      return `"${escapedString}"`;
+    }
+    if (typeof data === 'number' || typeof data === 'boolean' || data === null) {
+      return String(data);
+    }
+    if (validator.isArray(data)) {
+      const elements = data.map(item => this.objectToString(item)).join(', ');
+      return `[${elements}]`;
+    }
+    if (typeof data === 'object' && data !== null) {
+      // Filter out properties where the value is undefined BEFORE mapping
+      const kvPairs = Object.entries(data)
+        .filter(([, val]) => val !== undefined)
+        .map(([key, val]) => {
+          // GraphQL object keys are typically unquoted.
+          return `${key}: ${this.objectToString(val)}`;
+        });
+  
+      if (kvPairs.length === 0) {
+        return '{}'; // Represent an object with no defined properties as {}
+      }
+      return `{ ${kvPairs.join(', ')} }`;
+    }
+    
+    // If value is undefined (and not an object property, which is handled above,
+    // e.g., if objectToString(undefined) is called directly or for an array element)
+    // it should be represented as 'null'.
+    if (typeof data === 'undefined') {
+      return 'null';
+    }
+
+    // Fallback for any other types (e.g., Symbol, BigInt - though less common in GQL contexts)
+    // Consider how these should be handled or if an error should be thrown.
+    // For now, simple string conversion.
+    return String(data);
+  }
+
+  private formatTableName(tableName: string): string {
+    // Format tableName: first character to lowercase
+    if (tableName && tableName.length > 0) {
+      return tableName.charAt(0).toLowerCase() + tableName.slice(1);
+    }
+    return tableName;
+  }
+
+  private handleBulkImportErrors(err: FirebaseDataConnectError): never {
+    if (err.code === `data-connect/${DATA_CONNECT_ERROR_CODE_MAPPING.QUERY_ERROR}`){
+      throw new FirebaseDataConnectError(
+        DATA_CONNECT_ERROR_CODE_MAPPING.QUERY_ERROR,
+        `${err.message}. Make sure that your table name passed in matches the type name in your GraphQL schema file.`);
+    }
+    throw err;
+  }
+
+  /**
+   * Insert a single row into the specified table.
+   */
+  public async insert<GraphQlResponse, Variables extends object>(
+    tableName: string,
+    data: Variables,
+  ): Promise<ExecuteGraphqlResponse<GraphQlResponse>> {
+    if (!validator.isNonEmptyString(tableName)) {
+      throw new FirebaseDataConnectError(
+        DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT,
+        '`tableName` must be a non-empty string.');
+    }
+    if (validator.isArray(data)) {
+      throw new FirebaseDataConnectError(
+        DATA_CONNECT_ERROR_CODE_MAPPING.INVALID_ARGUMENT,
+        '`data` must be an object, not an array, for single insert.');
```

Comment:
I wonder if we should help even more and say `for arrays, please use "insertMany"`

Reply:
Good call. Updated!